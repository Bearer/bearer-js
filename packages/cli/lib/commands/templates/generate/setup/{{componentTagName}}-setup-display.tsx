/*
  The purpose of this component is to save scenario credentials.
*/

import { Component } from "@bearer/core";
import "@bearer/ui";

@Component({
  tag: "{{componentTagName}}-setup-display",
  shadow: true
})
export class {{scenarioTitle}}SetupDisplay {
  render() {
    return (
      <bearer-setup-display scenarioId="BEARER_SCENARIO_ID" />
    )
  }
}

