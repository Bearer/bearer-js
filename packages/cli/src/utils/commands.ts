import BaseCommand from '../base-command'

export async function linkIntegration(this: BaseCommand, anIdentifier: string) {
  let identifier = anIdentifier
  let { integrationTitle } = this.bearerConfig as { integrationTitle: string }
  if (!identifier) {
    const { data } = await this.devPortalClient.request<{ integrations: Integration[] }>({ query: QUERY })
    if (!data.data) {
      throw 'Unable to fetch integration list'
    }

    const { integrations } = data.data!
    const { integration } = await this.inquirer.prompt<{ integration: Integration }>([
      {
        name: 'integration',
        type: 'list',
        choices: integrations.map(int => ({
          name: `${int.uuid} - ${int.name}`,
          value: int
        }))
      }
    ])
    integrationTitle = integration.name
    identifier = integration.uuid
  }

  const integrationRc = { integrationTitle, integrationId: identifier }
  this.bearerConfig.setIntegrationConfig(integrationRc)
  this.log('Integration successfully linked! ðŸŽ‰')
}

type Integration = {
  uuid: string
  name: string
  latestActivity?: {
    state: string
  }
}

const QUERY = `
query CLILinkIntegrationList {
  integrations(includeGloballyAvailable: false) {
    uuid: uuidv2
    name
    latestActivity {
      state
    }
  }
}
`
