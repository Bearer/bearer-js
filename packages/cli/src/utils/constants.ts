import { BearerEnv, BaseConfig } from '../types'

export const LOGIN_CLIENT_ID = process.env.BEARER_LOGIN_CLIENT_ID || 'Wgll39KqWnJWud473wq7hZhiXxeNjEU7'
export const BEARER_ENV = process.env.BEARER_ENV || 'production'
export const BEARER_AUTH_PORT = 45677
export const BEARER_LOGIN_PORT = 56789

export const CONFIGS: Record<BearerEnv, BaseConfig> = {
  dev: {
    IntegrationServiceHost: 'https://int.dev.bearer.sh/',
    IntegrationServiceUrl: 'https://int.dev.bearer.sh/api/v1/',
    DeveloperPortalAPIUrl: 'https://api.staging.bearer.sh/graphql',
    DeveloperPortalUrl: 'https://app.staging.bearer.sh/',
    CdnHost: 'https://static.dev.bearer.sh',
    BearerEnv: 'dev',
    LoginDomain: 'https://login.bearer.sh'
  },
  staging: {
    IntegrationServiceHost: 'https://int.staging.bearer.sh/',
    IntegrationServiceUrl: 'https://int.staging.bearer.sh/api/v1/',
    DeveloperPortalAPIUrl: 'https://api.staging.bearer.sh/graphql',
    DeveloperPortalUrl: 'https://app.staging.bearer.sh/',
    CdnHost: 'https://static.staging.bearer.sh',
    BearerEnv: 'staging',
    LoginDomain: 'https://login.bearer.sh'
  },
  production: {
    IntegrationServiceHost: 'https://int.bearer.sh/',
    IntegrationServiceUrl: 'https://int.bearer.sh/api/v1/',
    DeveloperPortalAPIUrl: 'https://api.bearer.sh/graphql',
    DeveloperPortalUrl: 'https://app.bearer.sh/',
    CdnHost: 'https://static.bearer.sh',
    BearerEnv: 'production',
    LoginDomain: 'https://login.bearer.sh'
  }
}

// tslint:disable max-line-length
export const SUCCESS_LOGIN_PAGE = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Authentication callback</title>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style>
      * {
        -webkit-font-smoothing: antialiased;
        -moz-osx-font-smoothing: grayscale;
      }
      html,
      body {
        background-color: #f5f7fb;
        font-family: 'Proxima Nova', -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif,
          'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol';
        text-align: center;
        font-size: 16px;
        line-height: 16px;
      }
      h1 {
        color: #00c682;
        font-size: 2rem;
        font-weight: 600;
        letter-spacing: 0.99px;
        line-height: 29px;
      }
      p {
        color: #343c5d;
        letter-spacing: 0.56px;
      }
      a {
        border-radius: 4px;
        display: inline-block;
        margin-top: 32px;
        padding: 12px 18px;
        background-color: #030d36;
        color: #ffffff;
        font-weight: 600;
        letter-spacing: 0.2px;
        text-decoration: none;
      }
      .outer {
        display: table;
        position: absolute;
        top: 0;
        left: 0;
        height: 100%;
        width: 100%;
      }
      .middle {
        display: table-cell;
        vertical-align: middle;
      }
      .inner {
        margin-left: auto;
        margin-right: auto;
        max-width: 700px;
      }
      .hint {
        font-size: 0.9rem;
        padding: 1.5rem;
        border: 1px solid #c2c9ea;
        border-radius: 4px;
        background-color: #ffffff;
        position: relative;
        top: 60px;
        margin-bottom: 60px;
      }
    </style>
  </head>
  <body>
    <div class="outer">
      <div class="middle">
        <div class="inner">
          <svg width="37" height="40" xmlns="http://www.w3.org/2000/svg">
            <path
              d="M.8 19.2h25.269l-6.635-6.634a.8.8 0 0 1 1.132-1.132l8 8a.8.8 0 0 1 0 1.132l-8 8a.8.8 0 0 1-1.132-1.132L26.07 20.8H.8a.8.8 0 0 1 0-1.6zm14.4 11.2a.8.8 0 0 1 .8.8v6.4a.8.8 0 0 0 .8.8h17.6a.8.8 0 0 0 .8-.8V2.4a.8.8 0 0 0-.8-.8H16.8a.8.8 0 0 0-.8.8v6.4a.8.8 0 0 1-1.6 0V2.4A2.4 2.4 0 0 1 16.8 0h17.6a2.4 2.4 0 0 1 2.4 2.4v35.2a2.4 2.4 0 0 1-2.4 2.4H16.8a2.4 2.4 0 0 1-2.4-2.4v-6.4a.8.8 0 0 1 .8-.8z"
              fill="#00C682"
              fill-rule="nonzero"
            />
          </svg>
          <h1>Successfully authenticated</h1>
          <p>You can close this window</p>
          <br />
        </div>
      </div>
    </div>
  </body>
</html>
`
