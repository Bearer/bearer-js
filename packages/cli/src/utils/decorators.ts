import Command from '../BaseCommand'

type Constructor<T> = new (...args: any[]) => T
type TCommand = InstanceType<Constructor<Command>>

export function RequireScenarioFolder() {
  return function(_target: any, _propertyKey: string | symbol, descriptor: PropertyDescriptor) {
    let originalMethod = descriptor.value
    descriptor.value = async function(this: TCommand) {
      if (this.bearerConfig.isScenarioLocation) {
        await originalMethod.apply(this, arguments)
      } else {
        this.error('This command must be run within a scenario folder.')
      }
    }
    return descriptor
  }
}

export function RequireLinkedScenario() {
  return function(_target: any, _propertyKey: string | symbol, descriptor: PropertyDescriptor) {
    let originalMethod = descriptor.value
    descriptor.value = async function(this: TCommand) {
      if (this.bearerConfig.hasScenarioLinked) {
        await originalMethod.apply(this, arguments)
      } else {
        const error =
          this.colors.bold('You scenario must be linked before running this command\n') +
          this.colors.yellow(this.colors.italic('Please run: bearer link'))
        this.error(error)
      }
    }
    return descriptor
  }
}

export function ensureFreshToken() {
  return function(_target: any, _propertyKey: string | symbol, descriptor: PropertyDescriptor) {
    let originalMethod = descriptor.value
    descriptor.value = async function(this: TCommand) {
      const { authorization, ExpiresAt } = this.bearerConfig.bearerConfig

      if (authorization && authorization.AuthenticationResult) {
        try {
          if (ExpiresAt < Date.now()) {
            this.ux.action.start('Refreshing token')
            await refreshMyToken(this)
            this.ux.action.stop()
          }
        } catch (error) {
          this.ux.action.stop(`Failed`)
          this.error(error.message)
        }
        // TS is complaining with TS2445, commenting out this for now
        // this.debug('Running original method')
        await originalMethod.apply(this, arguments)
        return descriptor
      } else {
        const error =
          this.colors.bold('⚠️ It looks like you are not logged in\n') +
          this.colors.yellow(this.colors.italic('Please run: bearer login'))
        this.error(error)
        return descriptor
      }
    }
    return descriptor
  }
}

async function refreshMyToken(command: TCommand): Promise<boolean | Error> {
  const { RefreshToken } = command.bearerConfig.bearerConfig.authorization.AuthenticationResult!

  // try {
  const { statusCode, body } = await command.serviceClient.refresh({ RefreshToken })

  switch (statusCode) {
    case 200: {
      const {
        authorization: { AuthenticationResult }
      } = body

      await command.bearerConfig.storeBearerConfig({
        ExpiresAt: Date.now() + AuthenticationResult.ExpiresIn * 1000,
        authorization: {
          AuthenticationResult: {
            ...AuthenticationResult,
            RefreshToken
          }
        }
      })
      return true
    }
    case 401: {
      throw new UnauthorizedRefreshTokenError()
    }
    default: {
      throw new UnexpectedRefreshTokenError(JSON.stringify(body, undefined, 2))
    }
  }
}

class UnauthorizedRefreshTokenError extends Error {
  constructor() {
    super('Something went wrong, please run bearer login and try again')
  }
}

class UnexpectedRefreshTokenError extends Error {
  constructor(body: any) {
    super(`body : ${body}`)
  }
}
