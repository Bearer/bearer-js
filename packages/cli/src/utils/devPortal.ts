import axios from 'axios'
import BaseCommand from '../base-command'

export const devPortalClient = (command: BaseCommand) => {
  const instance = axios.create({
    baseURL: command.constants.DeveloperPortalAPIUrl
  })
  instance.interceptors.response.use(
    r => r,
    error => {
      if (error.response && error.response.status) {
        command.error('Unauthorized action, please run bearer login first')
      }
      return Promise.reject(error)
    }
  )

  return {
    request: async <DataReturned>(data: { query: string; variables?: any }) => {
      const token = await command.bearerConfig.getToken()
      if (token) {
        return instance.post<{ data?: DataReturned }>('', data, {
          headers: {
            Authorization: `Bearer ${token.access_token}`
          }
        })
      }
      throw new UnauthorizedError()
    }
  }
}

export class UnauthorizedError extends Error {
  constructor() {
    super('Unauthorized request')
  }
}
