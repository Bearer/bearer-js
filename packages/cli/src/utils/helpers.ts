import * as copy from 'copy-template-dir'
import * as path from 'path'

import BaseCommand from '../BaseCommand'

export function copyFiles(
  command: BaseCommand,
  sourceDirectory: string,
  outPutDirectory: string,
  vars: any
): Promise<boolean | string> {
  const base = process.cwd()
  function printFiles(files: Array<string>) {
    files.forEach(file => {
      command.log(command.colors.gray(`    create: `) + command.colors.white(file.replace(base + '/', '')))
    })
  }

  return new Promise((resolve, reject) => {
    copy(path.join(__dirname, '..', '..', 'templates', sourceDirectory), outPutDirectory, vars, (err, createdFiles) => {
      if (err) {
        reject(err)
      } else {
        printFiles(createdFiles)
        resolve(true)
      }
    })
  })
}
