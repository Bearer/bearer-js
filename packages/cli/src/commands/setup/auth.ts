import * as express from 'express'
import * as fs from 'fs'
import axios from 'axios'
import { modify, applyEdits } from 'jsonc-parser'
import { TConfig, Authentications, configs } from '@bearer/types/lib/authentications'
import { contexts } from '@bearer/functions/lib/declaration'

// TODO: use esModuleInterop config
// @ts-ignore
import * as open from 'open'

import BaseCommand from '../../base-command'
import { RequireIntegrationFolder } from '../../utils/decorators'
import { BEARER_AUTH_PORT, SUCCESS_LOGIN_PAGE } from '../../utils/constants'
import { askForString, askForPassword } from '../../utils/prompts'
import { startServer, TDestroyableServer, UnavailablePort } from '../../actions/startLocalServer'

type Event = 'success' | 'error'
const SEPARATOR = ':'
const COMMAND = `bearer setup:auth`
const enum keys {
  BEARER_AUTH_CLIENT_ID = 'BEARER_AUTH_CLIENT_ID',
  BEARER_AUTH_CLIENT_SECRET = 'BEARER_AUTH_CLIENT_SECRET',
  BEARER_AUTH_CONSUMER_KEY = 'BEARER_AUTH_CONSUMER_KEY',
  BEARER_AUTH_CONSUMER_SECRET = 'BEARER_AUTH_CONSUMER_SECRET',
  BEARER_AUTH_USERNAME = 'BEARER_AUTH_USERNAME',
  BEARER_AUTH_PASSWORD = 'BEARER_AUTH_PASSWORD',
  BEARER_AUTH_APIKEY = 'BEARER_AUTH_APIKEY'
}
export default class SetupAuth extends BaseCommand {
  static description = `setup API credentials for local development.
If you would like to bypass the prompt, you can either:
\t* pass credentials as argument (see description later)
\t* use environment variables
see examples`

  static flags = {
    ...BaseCommand.flags
  }

  static examples = [
    `With argument`,
    `\t${COMMAND} CLIENT_ID${SEPARATOR}CLIENT_SECRET`,
    `\t${COMMAND} CONSUMER_KEY${SEPARATOR}CONSUMER_SECRET`,
    `\t${COMMAND} USERNAME${SEPARATOR}PASSWORD`,
    `\t${COMMAND} APIKEY`,
    `With environment variables`,
    `\t${keys.BEARER_AUTH_CLIENT_ID}=CLIENT_ID ${keys.BEARER_AUTH_CLIENT_SECRET}=CLIENT_SECRET ${COMMAND}`,
    `\t${keys.BEARER_AUTH_CONSUMER_KEY}=CONSUMER_KEY ${keys.BEARER_AUTH_CONSUMER_SECRET}=CONSUMER_SECRET ${COMMAND}`,
    `\t${keys.BEARER_AUTH_USERNAME}=USERNAME ${keys.BEARER_AUTH_CONSUMER_SECRET}=PASSWORD ${COMMAND}`,
    `\t${keys.BEARER_AUTH_APIKEY}=APIKEY ${COMMAND}`
  ]

  static args = [
    {
      name: 'credentials',
      description: `Provide inline credentials`,
      required: false,
      default: ''
    }
  ]

  _server?: TDestroyableServer
  _verifier!: string
  _challenge!: string
  private _listerners: {
    success: ((data: TBase64EncodedString) => void)[]
    error: ((data: any) => void)[]
  } = {
    success: [],
    error: []
  }

  @RequireIntegrationFolder()
  async run() {
    const { args } = this.parse(SetupAuth)
    const config: TConfig = JSON.parse(
      fs.readFileSync(this.locator.authConfigPath, {
        encoding: 'utf8'
      })
    )
    const { authType, provider, hint } = config as { authType: Authentications; provider?: string; hint?: string }

    if (hint) {
      this.log(this.colors.italic(this.colors.yellow(`\n${hint}\n`)))
    }

    function prefixedPrompt(message: string) {
      return prefix(message, provider)
    }

    switch (authType) {
      case Authentications.OAuth2: {
        const [idArg, secretArg] = args.credentials.split(SEPARATOR)
        const {
          [keys.BEARER_AUTH_CLIENT_ID]: id = idArg,
          [keys.BEARER_AUTH_CLIENT_SECRET]: secret = secretArg
        } = process.env
        const clientID = id || (await askForString(prefixedPrompt('Client ID'), { type: 'password' }))
        const clientSecret = secret || (await askForString(prefixedPrompt('Client secret'), { type: 'password' }))

        this.debug('Your credentials:\n%j', { ...config, clientID, clientSecret: clientSecret.replace(/./g, '*') })

        const newConfig = { ...config, clientID, clientSecret }
        const token = await this.fetchAuthToken(newConfig as configs.TOAuth2Config)
        const setup = JSON.parse(Buffer.from(token, 'base64').toString('ascii')) as contexts.OAuth2
        await this.persistSetup(setup)
        break
      }
      case Authentications.OAuth1: {
        const [keyArg, secretArg] = args.credentials.split(SEPARATOR)
        const {
          [keys.BEARER_AUTH_CONSUMER_KEY]: key = keyArg,
          [keys.BEARER_AUTH_CONSUMER_SECRET]: secret = secretArg
        } = process.env

        const consumerKey = key || (await askForString(prefixedPrompt('Consumer key')))
        const consumerSecret = secret || (await askForPassword(prefixedPrompt('Consumer secret')))

        this.debug('Your credentials:\n%j', { consumerKey, consumerSecret: consumerSecret.replace(/./g, '*') })

        const newConfig = { ...config, consumerKey, consumerSecret }
        const token = await this.fetchAuthToken(newConfig as configs.TOAuth1Config)
        const setup = JSON.parse(Buffer.from(token, 'base64').toString('ascii')) as contexts.OAuth1
        // TODO: fix this when context.OAuth1 is fixed  and well defined
        await this.persistSetup(setup)
        break
      }

      case Authentications.Basic: {
        const [usernameArg, passwordArg] = args.credentials.split(SEPARATOR)
        const {
          [keys.BEARER_AUTH_USERNAME]: basicUsername = usernameArg,
          [keys.BEARER_AUTH_PASSWORD]: basicPassword = passwordArg
        } = process.env
        const username = basicUsername || (await askForString(prefixedPrompt('Username')))
        const password = basicPassword || (await askForPassword(prefixedPrompt('Password')))

        await this.persistSetup({ username, password } as contexts.Basic)

        break
      }
      case Authentications.ApiKey: {
        const { [keys.BEARER_AUTH_APIKEY]: key } = process.env
        const apiKey = key || args.credentials || (await askForPassword(prefixedPrompt('API Key')))
        await this.persistSetup({ apiKey } as contexts.ApiKey)
        break
      }
      case Authentications.Custom:
      case Authentications.NoAuth: {
        await this.persistSetup({})
        return this.warn(
          `The current authentication type of this integration is not supported by this command: ${authType}`
        )
      }
      default:
        // unsure we handled all authentications
        // http://ideasintosoftware.com/exhaustive-switch-in-typescript/
        this.error(`The current authentication type of this integration is not supported by this command: ${authType}`)
        throw new UnreachableCaseError(authType)
    }
  }

  async fetchAuthToken(config: configs.TOAuth2Config | configs.TOAuth1Config): Promise<TBase64EncodedString> {
    return new Promise(async (resolve, reject) => {
      this._server = await this.startServer()
      const redirectLocation = await axios
        .post(`${this.constants.IntegrationServiceHost}v2/auth/local-auth`, { config }, { maxRedirects: 0 })
        .catch(e => e.response.headers.location)

      // add listeners so that we can handle redirect and fallback the same way
      this.on<TBase64EncodedString>('success', data => {
        resolve(data)
      })

      this.on('error', data => {
        this.debug(data)
        reject('Error while receiving token')
      })

      const { url, fallback } = getOpeningUrls(
        `${this.constants.IntegrationServiceHost}v2/auth/${redirectLocation.replace('./', '')}&clientId=NONE`
      )
      this.debug('config: %j, location: %s', config, url)
      const a = await open(url)
      a.on('close', (code: any, signal: any) => {
        if (code !== 0) {
          // we are no able to open a browser so we ask the user to fill with the token
          this.stopServer()
          this.warn(
            this.colors.yellow(`Unable to open a browser. If you want to retrieve a token please follow these steps\n`)
          )
          this.log(this.colors.bold('1/ access the url below and follow the login process:\n\n'), fallback)
          this.log()
          this.log(this.colors.bold(`2/ when you access the success page copy the token and paste it here`))
          askForString('Token').then((token: TBase64EncodedString) => {
            resolve(token)
          })
        }
      })
    })
  }

  private stopServer = () => {
    return new Promise((resolve, _reject) => {
      if (this._server) {
        this.debug('stopping server')
        this._server.destroy(() => {
          this.debug('server stopped')
          resolve()
        })
      } else {
        resolve()
      }
    })
  }

  private startServer = async (): Promise<TDestroyableServer> => {
    this.on('success', this.stopServer)
    this.on('error', this.stopServer)

    try {
      const app = express()

      app.use((_req: express.Request, res: express.Response, next: express.NextFunction) => {
        res.setHeader('Connection', 'close')
        next()
      })

      app.get('/setup/auth-callback', (req: express.Request, res: express.Response) => {
        const token: TBase64EncodedString = req.query.token || ''
        try {
          res.send(SUCCESS_LOGIN_PAGE).end()
          setTimeout(() => {
            this._listerners.success.map(cb => cb(token))
          }, 0)
        } catch (e) {
          this.debug(e)
          this._listerners.error.map(cb => cb(e))
          this.error('Error while handling callback')
        }
      })

      this.debug('starting server')
      return await startServer(BEARER_AUTH_PORT, app)
    } catch (e) {
      if (e instanceof UnavailablePort) {
        this.error('bearer setup:auth cannot start')
      }
      throw e
    }
  }

  private persistSetup(config: contexts.OAuth2 | contexts.OAuth1 | contexts.ApiKey | contexts.Basic) {
    const setupRc = this.locator.localConfigPath
    if (!fs.existsSync(setupRc)) {
      fs.writeFileSync(setupRc, '{}', { encoding: 'utf8' })
    }
    const rawSetup = fs.readFileSync(setupRc, { encoding: 'utf8' })
    const updates = modify(rawSetup, ['setup', 'auth'], config, {
      formattingOptions: {
        insertSpaces: true,
        tabSize: 2,
        eol: '\n'
      }
    })
    const newSetupRc = applyEdits(rawSetup, updates)
    this.debug('Writing setup config\n%j', { config, setupRc: newSetupRc })
    fs.writeFileSync(setupRc, newSetupRc, { encoding: 'utf8' })
    this.success(`Auth credentials have been saved to ${this.locator.toRelative(setupRc)}`)
  }

  private on = <T>(event: Event, callback: (data: T) => void) => {
    this._listerners[event] = [...this._listerners[event], callback as any]
  }
}

function getOpeningUrls(url: string) {
  return {
    url: `${url}&localHostRedirectSupported=true`,
    fallback: `${url}&localHostRedirectSupported=inline`
  }
}

function prefix(message: string, prefix?: string) {
  const prefixedMessage = [prefix, message].filter(Boolean).join(' ')

  return `Enter ${prefixedMessage}`
}

class UnreachableCaseError extends Error {
  constructor(val: never) {
    super(`Unreachable case: ${val}`)
  }
}

type TBase64EncodedString = string
