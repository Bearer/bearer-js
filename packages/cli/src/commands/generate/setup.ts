import { flags } from '@oclif/command'
import * as fs from 'fs'
import * as path from 'path'

import BaseCommand from '../../BaseCommand'
import { RequireScenarioFolder } from '../../utils/decorators'
import { copyFiles } from '../../utils/helpers'

export default class GenerateSetup extends BaseCommand {
  static description = 'Generate a Bearer Setup'
  static hidden = true
  static flags = {
    ...BaseCommand.flags,
    help: flags.help({ char: 'h' }),
    force: flags.boolean({})
  }

  static args = []

  @RequireScenarioFolder()
  async run() {
    const { flags } = this.parse(GenerateSetup)
    if (flags.force || !setupExists(this.locator.srcViewsDir)) {
      const fields = this.scenarioAuthConfig.setupViews
      if (fields && fields.length) {
        try {
          await copyFiles(
            this,
            'generate/setup',
            this.locator.srcViewsDir,
            this.getVars(this.bearerConfig.scenarioConfig.scenarioTitle, fields)
          )
          this.success('Setup components successfully generated! ðŸŽ‰')
        } catch (e) {
          this.error(e)
        }
      } else {
        this.warn(`No setupViews key found within auth.config.json file. skipping`)
      }
    } else {
      this.warn('Setup files already exists, use --force flag to overwrite existing ones.')
    }
  }

  getVars(scenarioTitle: string, fields: any) {
    return {
      componentName: this.case.pascal(scenarioTitle),
      componentTagName: this.case.kebab(scenarioTitle),
      fields: JSON.stringify(fields)
    }
  }
}

// Note: using Or condition in case the developer delete one but customized the other component
function setupExists(location: string): boolean {
  return (
    fs.existsSync(path.join(location, 'setup-action.tsx')) || fs.existsSync(path.join(location, 'setup-display.tsx'))
  )
}
