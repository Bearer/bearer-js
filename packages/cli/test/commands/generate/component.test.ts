import GenerateComponent from '../../../src/commands/generate/component'
import { ensureBearerStructure } from '../../helpers/setup'
import { readFile } from '../../helpers/utils'

describe('Generate', () => {
  let bearerPath: string
  let result: string[]

  beforeEach(() => {
    result = []
    jest.spyOn(process.stdout, 'write').mockImplementation(val => {
      result.push(val)
      return true
    })
    bearerPath = ensureBearerStructure({ withViews: true })
  })

  describe('generate:component', () => {
    it('blank component', async () => {
      await GenerateComponent.run(['blankComponent', '-t', 'blank', '--path', bearerPath])
      expect(result.join()).toContain('Component generated')
      expect(readFile(bearerPath, 'views', 'components', 'blankComponent.tsx')).toMatchSnapshot()
    })

    it('collection component', async () => {
      await GenerateComponent.run(['collectionComponent', '-t', 'collection', '--path', bearerPath])
      expect(result.join()).toContain('Component generated')
      expect(readFile(bearerPath, 'views', 'components', 'collectionComponent.tsx')).toMatchSnapshot()
    })

    it('root component', async () => {
      await GenerateComponent.run(['rootComponent', '-t', 'root', '--path', bearerPath])
      expect(result.join()).toContain('Component generated')
      expect(readFile(bearerPath, 'views', 'root-component.tsx')).toMatchSnapshot()
    })
  })
})
