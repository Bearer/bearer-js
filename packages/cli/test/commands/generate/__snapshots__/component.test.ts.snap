// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate generate:component blank component 1`] = `
"import { Component } from '@bearer/core'

@Component({
  tag: 'blank-component',
  styleUrl: 'blankComponent.css',
  shadow: true
})
export class BlankComponent {
  render() {
    return (
      <div class=\\"root\\">
        <bearer-typography as=\\"h2\\">
          BlankComponent Component
        </bearer-typography>
      </div>
    )
  }
}
"
`;

exports[`Generate generate:component collection component 1`] = `
"import { BearerFetch, Component, Intent } from '@bearer/core'

@Component({
  tag: 'collection-component',
  shadow: true
})
export class CollectionComponent {
  @Intent('collectionComponent') fetcher: BearerFetch
  render() {
    return <bearer-scrollable fetcher={this.fetcher} />
  }
}"
`;

exports[`Generate generate:component root component 1`] = `
"/*
  The purpose of this component is to deal with scenario navigation between each views.

*/

import { RootComponent } from '@bearer/core'
import '@bearer/ui'

@RootComponent({
  role: 'action',
  group: 'root-component'
})
export class RootComponentAction {
  render() {
    return (
      <bearer-navigator btnProps={ { content:\\"RootComponent Action\\", kind:\\"primary\\" } } direction=\\"right\\">
        
        <bearer-navigator-screen navigationTitle=\\"My first screen\\">
          <div style={ { textAlign: 'center' } }>My first screen</div>
        </bearer-navigator-screen>
      </bearer-navigator>
    )
  }
}
"
`;

exports[`Generate generate:component root component 2`] = `
"/*
  The purpose of this component is to be the result of your scenario.
  Its responsibility is to retrieve the scenario state from a previous action
  of a user.
*/
import { RootComponent } from '@bearer/core'
import '@bearer/ui'

@RootComponent({
  role: 'display',
  group: 'root-component'
})
export class RootComponentDisplay {
  render() {
    return <bearer-alert kind=\\"success\\">ðŸš€ My RootComponent Display component ðŸš€</bearer-alert>
  }
}
"
`;
